File: lua/features/mine.lua
local x -- Columns
local y -- Rows

local map = {} -- create maps
x = 7
y = 7

-- init map by setting it blank
for i = 1,y do
	map[i] = map[i] or {}
	for j = 1,x do
		map[i][j] = { -- tile object
			value = 0;			-- default tile value
			checked = false;	-- default tile checked
			flaged = false;		-- default tile flaged
		}
	end
end

function drawBoard() -- draw the map
	str = "" -- blank string
	
	-- for each row and column
	for i = 1,y do
		for j = 1,x do
			-- append the value of the tile
			if not(map[i][j].checked) then
				str = str.." x,"
			else
				if (map[i][j].value < 0) then
					str = str..map[i][j].value..","
				else
					str = str.." "..map[i][j].value..","
				end
			end
		end
		
		-- newline
		str = str.."\n"
	end
	
	-- print to screen
	print(str)
end

function getPlayerInput(input)end

function checkTile(tile) -- check if tile is a bomb
	-- is tile less than 0
	if(tile.value < 0) then
		return true
	end
	return false
end

--//-------New-CHANGE------------//--
function revealtile(tile)
	tile.checked = true
end
--//----------------------------//--

function mapClear()
	ret = false
	
	-- for each row and column
	for i = 1,y do
		for j = 1,x do
			if not(checkTile(map[i][j])) then
				if not(map[i][j].checked) then
					ret = true
				end
			end
		end
	end
	
	return ret
end

--//-------New-CHANGE------------//--
function placemines()
	-- for each row and column
	for i = 1,y do
		for j = 1,x do
			if((math.random(1,10) % 2) == 1) then
				-- set bombs randomly
				map[i][j].value = -1
			end
		end
	end
end

function showAll()
	-- for each row and column
	for i = 1,y do
		for j = 1,x do
			map[i][j].checked = true
		end
	end
end
--//----------------------------//--
--// removed: print("x:"..x..", y:"..y)
--// removed: drawBoard()

--//-------New-CHANGE------------//--
local sx,sy
placemines(map)	-- place mines
repeat
	print("\n")
	drawBoard()	-- draw the map
	
	-- get player input
	print("enter tile x:")				-- input text for x
	sx = io.read(); sx = tonumber(sx)	-- read x from stdin
	print("enter tile y:")				-- input text for y
	sy = io.read(); sy = tonumber(sy)	-- read y from stdin
	
	if(checkTile(map[sy][sx])) then
		-- if tile is a bomb
		print("\nTHAT'S A BOMB");
		print("GAME OVER\n");	-- loser XD
		break;
	end
	
	revealtile(map[sy][sx])	-- show the tile
until not(mapClear())	-- repeat until map is not clear
if (mapClear()) then
	print("\nYOU WIN\n")	-- winner :D
end
showAll()		-- show the map
drawBoard()		-- draw the map
--//----------------------------//--
return 0; -- Zero Errors

File: java/features/mine.java
//-------New-CHANGE------------//
import java.util.Scanner;
//----------------------------//

class tile {
	private int value;			// tile value
	private boolean checked;	// tile checked
	private boolean flaged;		// tile flaged
	
	tile(int v){
		value = v;			// set value
		checked = false;	// defualt
		flaged = false;		// defualt
	}
	
	public int getVal(){ // return tile value
		return this.value;
	}
	
	public boolean getCheck(){ // return tile checked
		return this.checked;
	}
	
	public boolean getflag(){ // return tile flaged
		return this.flaged;
	}
	
	//-------New-CHANGE------------//
	public void setVal(int in){ // set tile value
		this.value = in;
	}
	
	public void setCheck(boolean in){ // set tile checked
		this.checked = in;
	}
	
	public void setflag(boolean in){ // set tile flaged
		this.flaged = in;
	}
	//----------------------------//
}

class map {
	int x;			// Columns
	int y;			// Rows
	tile[][] Map;	// Tile map
	
	map(){ // no input method
		x = 7;					// defualt size
		y = 7;					// defualt size
		Map = new tile[y][x];	// defualt size
		
		// for every row and column in the Map
		for(int i = 0; i < y; i++){
			for(int j = 0; j < x; j++){
				// Set tiles
				Map[i][j] = new tile(0);
			} 
		}
	}
	
	map(int ix, int iy){ // custom input method
		if(ix > 0){ x = ix; }else{ x = 7; }	// valid size or defualt
		if(iy > 0){ y = iy; }else{ y = 7; } // valid size or defualt
		Map = new tile[iy][ix];				// custom map size
		
		// for every row and column in the Map
		for(int i = 0; i < iy; i++){
			for(int j = 0; j < ix; j++){
				// Set tiles
				Map[i][j] = new tile(0);
			} 
		}
	}
	
	//-------New-CHANGE------------//
	public void placemines(){
		for(int i = 0; i < this.y; i++){
			for(int j = 0; j < this.x; j++){
				// plant bomb randomly tile
				if(((int)((Math.random() * 10) % 2)) == 1){
					Map[i][j] = new tile(-1);
				}
			} 
		}
	}
	//----------------------------//
	
	public void drawBoard(){
		String ret = "";	// create blank string
		
		// for every row and column in the Map
		for(int i = 0; i < this.y; i++){
			for(int j = 0; j < this.x; j++){
				// add value to string
				//-------New-CHANGE------------//
				if(!this.Map[i][j].getCheck()){
					ret = ret + " x,";
				}else{
					if(this.Map[i][j].getVal() < 0){
						ret = ret + this.Map[i][j].getVal() + ",";
					}else{
						ret = ret + " " + this.Map[i][j].getVal() + ",";
					}
				}
				//----------------------------//
			}
			
			// newline
			ret = ret + "\n";
		}
		
		// print string
		System.out.println(ret);
	}
	
	public boolean checkTile(int x,int y){ // Check tile at (x,y)
		// check if value is bomb
		if(this.Map[y][x].getVal() < 0){ return true; }
		return false;
	}
	
	//-------New-CHANGE------------//
	public void revealtile(int x, int y){
		this.Map[y][x].setCheck(true);
	}
	
	public boolean mapClear(){ // Check if the map is clear
		boolean ret = false;
		
		// for every row and column in the Map
		for(int i = 0; i < this.y; i++){
			for(int j = 0; j < this.x; j++){
				if(!this.checkTile(j,i)){
					if(!this.Map[j][i].getCheck()){
						ret = true;
					}
				}
			}
		}
		
		return ret;
	}
	
	public void showAll(){
		// for every row and column in the Map
		for(int i = 0; i < this.y; i++){
			for(int j = 0; j < this.x; j++){
				// show the tile
				this.Map[j][i].setCheck(true);
			}
		}
	}
	//----------------------------//
}

public class mine{ // Public Class
	public static void main(String[] args){ // Starter Function
		//-------New-CHANGE------------//
		Scanner in = new Scanner(System.in);
		int sx,sy;
		//----------------------------//
		map M = new map(); // create map
		// removed: System.out.println("x:" + M.x + ", y:" + M.y);
		// removed: M.drawBoard(); // Draw The Board
		//-------New-CHANGE------------//
		M.placemines(); // place mines
		do{
			System.out.println("");
			M.drawBoard(); // draw the map
			
			// get player input
			System.out.println("enter tile x: ");	// input text for x
			sx = in.nextInt();						// read x from stdin
			System.out.println("enter tile y: ");	// input text for y
			sy = in.nextInt();						// read y from stdin
			
			if(M.checkTile(sx - 1,sy - 1)){
				// if tile is a bomb
				System.out.println("\nTHAT'S A BOMB");
				System.out.println("GAME OVER\n");	// loser XD
				break;
			}
			
			M.revealtile(sx - 1,sy - 1); // show the tile
		}while(M.mapClear()); // while the map is not clear
		if(!M.mapClear()){
			System.out.println("\nYOU WIN\n"); // winner :D
		}
		M.showAll();	// show the map
		M.drawBoard();	// draw the map
		//----------------------------//
		// Zero Errors
	}
}
